cmd/clusterctl/client/cluster/topology_plan_test.go

```diff
@@ -0,0 +1,492 @@
+					return []client.ObjectKey{cluster, cluster2}
+				}(),
+				affectedClusterClasses: func() []client.ObjectKey {
+					cc := client.ObjectKey{Namespace: "default", Name: "my-cluster-class"}
+					return []client.ObjectKey{cc}
+				}(),
+				modified: []item{
+					{kind: "KubeadmControlPlane", namespace: "default", namePrefix: "my-cluster-"},
+				},
+				created: []item{
+					// Modifying the DockerClusterTemplate will result in template rotation. A new template will be created
+					// and used by KCP.
+					{kind: "DockerMachineTemplate", namespace: "default", namePrefix: "my-cluster-control-plane-"},
+				},
+				reconciledCluster: &client.ObjectKey{Namespace: "default", Name: "my-cluster"},
+			},
+			wantErr: false,
+		},
+		{
+			name:            "Input with objects in different namespaces should return error",
+			existingObjects: []*unstructured.Unstructured{},
+			args: args{
+				in: func() *TopologyPlanInput {
+					objs, err := utilyaml.ToUnstructured(objsInDifferentNamespacesYAML)
+					if err != nil {
+						panic(err)
+					}
+					return &TopologyPlanInput{
+						Objs: convertToPtrSlice(objs),
+					}
+				}(),
+			},
+			wantErr: true,
+		},
+		{
+			name: "Input with TargetNamespace different from objects in input should return error",
+			args: args{
+				in: func() *TopologyPlanInput {
+					objs, err := utilyaml.ToUnstructured(newClusterClassAndClusterYAML)
+					if err != nil {
+						panic(err)
+					}
+					return &TopologyPlanInput{
+						Objs:            convertToPtrSlice(objs),
+						TargetNamespace: "different-namespace",
+					}
+				}(),
+			},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+
+			existingObjects := []client.Object{}
+			for _, o := range tt.existingObjects {
+				existingObjects = append(existingObjects, o)
+			}
+			proxy := test.NewFakeProxy().WithClusterAvailable(true).WithFakeCAPISetup().WithObjs(existingObjects...)
+			inventoryClient := newInventoryClient(proxy, nil)
+			tc := newTopologyClient(
+				proxy,
+				inventoryClient,
+			)
+
+			res, err := tc.Plan(tt.args.in)
+			if tt.wantErr {
+				g.Expect(err).To(HaveOccurred())
+				return
+			}
+			g.Expect(err).NotTo(HaveOccurred())
+			g.Expect(len(res.ClusterClasses)).To(Equal(len(tt.want.affectedClusterClasses)))
+			for _, cc := range tt.want.affectedClusterClasses {
+				g.Expect(res.ClusterClasses).To(ContainElement(cc))
+			}
+			g.Expect(len(res.Clusters)).To(Equal(len(tt.want.affectedClusters)))
+			for _, cluster := range tt.want.affectedClusters {
+				g.Expect(res.Clusters).To(ContainElement(cluster))
+			}
+			if tt.want.reconciledCluster == nil {
+				g.Expect(res.ReconciledCluster).To(BeNil())
+			} else {
+				g.Expect(res.ReconciledCluster).NotTo(BeNil())
+				g.Expect(*res.ReconciledCluster).To(Equal(*tt.want.reconciledCluster))
+			}
+			for _, created := range tt.want.created {
```

Comment:
Q: do we have to compare lengths? I think this wouldn't catch if res.Created, actualModifiedObjs and res.Deleted contain more entries

Reply:
Because of the way the tests are setup (loading objects from yamls to a fake object tracker) I observed that there could be more modified objects than actual. This way of testing makes sure that all the expected objects are in the real output. This way the tests will capture if an object was expected to be modified (lets say) but was not. However, there could be more objects in the output than expected.

Reply:
Okay got it, thx for the explanation!